oo-accept-node: Test BROKER_HOST & CLOUD_DOMAIN are consistent

dinhxuanvu opened the pull request:
The oo-diagnostics doesn't compare SSL certs from broker & BROKER_HOST
to see if they match since it only collects SSL cert from broker only.

After the commit, the oo-diagnostics gets SSL cert from BROKER_HOST along
with current broker. Then, it compares issuer, subject and common name
records from SSL certs to see if they match. If not, errors are issued.

This commit fixes bug 1054441

Signed-off-by: Vu Dinh vdinh@redhat.com

Miciah commented on node-util/sbin/oo-accept-node:
```diff
@@ -337,6 +337,23 @@ def check_node_public_resolution
+  # check the configuration variables (BROKER_HOST & CLOUD_DOMAIN) against env variables
+  begin
+    broker_host_env = File.open("#{$CONF_DIR}/env/OPENSHIFT_BROKER_HOST", "r").read
+    cloud_domain_env = File.open("#{$CONF_DIR}/env/OPENSHIFT_CLOUD_DOMAIN", "r").read
```
Better to use `File.read` because it's shorter and simpler and will close the FD for you.

tiwillia commented:
The file descriptors need to be closed, or `File.read` should be used as Miciah suggested.

Otherwise, LGTM.

dinhxuanvu committed:
oo-diagnostics: SSL cert matching for broker & BROKER_HOST

The oo-diagnostics doesn't compare SSL certs from broker & BROKER_HOST
to see if they match since it only collects SSL cert from broker only.

After the commit, the oo-diagnostics gets SSL cert from BROKER_HOST along
with current broker. Then, it compares issuer, subject and common name
records from SSL certs to see if they match. If not, errors are issued.

Bug <1054441>
Link <https://bugzilla.redhat.com/show_bug.cgi?id=1054441>

Signed-off-by: Vu Dinh <vdinh@redhat.com>

dinhxuanvu commented:
I fixed the File.open to use File.read instead as suggested. Also, I have added the commit for oo-diagnostics fix.

tiwillia commented:
[test]

tiwillia commented on common/bin/oo-diagnostics:
```diff
@@ -1226,6 +1228,28 @@ class OODiag
+        An error arose while trying to connect to the SSL reverse proxy.  curl returned with
+        #{env_response.strip.empty? ? "no output." : "the following output:\n" + env_response}
+
+        Please check whether the httpd service is running.
```
In this error output, perhaps show that we attempted to connect to `#{broker_host_env}` and failed. Otherwise, the errors for `curl`ing localhost and `#{broker_host_env}` are the same.

tiwillia commented on common/bin/oo-diagnostics:
```diff
@@ -1226,6 +1228,28 @@ class OODiag
+    # does ssl cert from broker_host_env matching the current broker
+    do_fail("SSL cert issuer doesn't match between current broker and broker host") if (env_issuer != issuer)
+    do_fail("SSL cert subject doesn't match between current broker and broker host") if (env_subject != subject)
+    do_fail("SSL cert commonname doesn't match between current broker and broker host") if (env_commonname != commonname)
```
These error messages are likely not very helpful to an administrator. The admin would not know the difference between 'current broker' and 'broker host'.

Perhaps something along the lines of:
```
do_fail("SSL cert issuer doesn't match between localhost and #{broker_host_env} defined in /etc/openshift/env/OPENSHIFT_BROKER_HOST")
```

dinhxuanvu commented:
@tiwillia Fixed the error messages to be more specific as requested.

tiwillia commented on common/bin/oo-diagnostics:
```diff
@@ -1207,14 +1207,16 @@ class OODiag
         #{response.strip.empty? ? "no output." : "the following output:\n" + response}
 
-        Please check whether the httpd service is running.
+        Please check whether the httpd service in #{broker_host_env} is running
```
Looks like you made the change in the output for localhost, but meant to make the change in the output for `#{broker_host_env}`

dinhxuanvu commented:
@tiwillia Fixed the mistake. Sorry about that!

tiwillia commented:
LGTM assuming tests pass.

Miciah commented on common/bin/oo-diagnostics:
```diff
@@ -1226,6 +1228,28 @@ class OODiag
+    end
+
+    # Get ssl cert's issuer, subject and commonname
+    env_issuer = env_response.slice(/^\*\s*issuer: .*$/).gsub(/^\*\s*issuer: /,"")
```
Might be more elegant to use `env_response.scan(/^\*\s*issuer: (.*)$/).first` or `/^\*\s*issuer: (.*)$/.match(env_response).first`.

Miciah replied:
Oh, I see we already use `.slice(//).gsub(//)` above, so never mind.

Miciah commented on common/bin/oo-diagnostics:
```diff
@@ -1226,6 +1228,28 @@ class OODiag
+    env_subject = env_response.slice(/^\*\s*subject: .*$/).gsub(/^\*\s*subject: /,"")
+    env_commonname = env_response.slice(/^\*\s*common name: .*$/).gsub(/^\*\s*common name: /,"")
+
+    # does ssl cert from broker_host_env matching the current broker
```
s/matching/match/

It's good to use full sentences with capitalisation and punctuation, but we don't always do that, so I guess it's not a big deal.

Miciah commented on common/bin/oo-diagnostics:
```diff
@@ -1226,6 +1228,28 @@ class OODiag
+    env_commonname = env_response.slice(/^\*\s*common name: .*$/).gsub(/^\*\s*common name: /,"")
+
+    # does ssl cert from broker_host_env matching the current broker
+    do_fail("SSL cert issuer doesn't match between localhost and #{broker_host_env} defined in /etc/openshift/env/OPENSHIFT_BROKER_HOST") if (env_issuer != issuer)
```
You don't need the parentheses in the `if` condition (or in the `do_fail` call, for that matter), but they don't hurt anything.

Miciah commented on common/bin/oo-diagnostics:
```diff
@@ -1207,6 +1207,8 @@ class OODiag
   def test_broker_certificate
     skip_test unless @is_broker
 
+    broker_host_env = File.read("/etc/openshift/env/OPENSHIFT_BROKER_HOST")
```
Might as well move this closer to where it's used.

dinhxuanvu commented:
@tiwillia @Miciah Fixed minor issues from Miciah.

dinhxuanvu commented:
@tiwillia @Miciah The testing is good. If you guys are not having any more concerns, please go ahead and merge whenever you can. Thanks.

dobbymoodge commented:
maybe [merge] ?

dinhxuanvu commented:
@tiwillia @dobbymoodge @Miciah If there is no other concern, please merge it when you can. Thanks.

Miciah commented:
[merge] please!

Miciah commented:
Looks like the merge attempt was foiled by EC2 timeouts.  Please try [merge]ing again!
